spec/lang/representation.md

```diff
@@ -405,22 +402,157 @@ Unsized types do not have values and thus there is no representation relation.
+                    // (Think: uninhabited types have impossible alignment.)
+                    ensure_else_ub(ptr.thin_pointer.addr != 0, "Value::Ptr: null safe pointer")?;
+                    ensure_else_ub(layout.align.is_aligned(ptr.thin_pointer.addr), "Value::Ptr: unaligned safe pointer")?;
+                    ensure_else_ub(layout.inhabited, "Value::Ptr: safe pointer to uninhabited type")?;
+                    ensure_else_ub(
+                        self.mem.dereferenceable(ptr.thin_pointer, size).is_ok(),
+                        "Value::Ptr: non-dereferenceable safe pointer"
+                    )?;
+
+                    // However, we do not care about the data stored wherever this pointer points to.
+                }
+
+                match ptr.metadata {
+                    Some(PointerMeta::ElementCount(num)) =>
+                        self.check_value(Value::Int(num), Type::Int(IntType::usize_ty::<M::T>()))?,
+                    _ => {}
+                };
+            }
+            (Value::Tuple(vals), Type::Tuple { fields, .. }) => {
+                ensure_else_ub(vals.len() == fields.len(), "Value::Tuple: invalid number of fields")?;
+                for (val, (_, ty)) in vals.zip(fields) {
+                    self.check_value(val, ty)?;
+                }
+            }
+            (Value::Tuple(vals), Type::Array { elem, count }) => {
+                ensure_else_ub(vals.len() == count, "Value::Tuple: invalid number of elements")?;
+                for val in vals {
+                    self.check_value(val, elem)?;
+                }
+            }
+            (Value::Union(chunk_data), Type::Union { chunks, .. }) => {
+                ensure_else_ub(chunk_data.len() == chunks.len(), "Value::Union: invalid chunk size")?;
+                for (data, (_, size)) in chunk_data.zip(chunks) {
+                    ensure_else_ub(data.len() == size.bytes(), "Value::Union: invalid chunk data")?;
+                }
+            }
+            (Value::Variant { discriminant, data }, Type::Enum { variants, .. }) => {
+                let Some(variant) = variants.get(discriminant) else {
+                    throw_ub!("Value::Variant: invalid discrimant");
+                };
+                self.check_value(data, variant.ty)?;
+            }
+            (_, Type::Slice { .. }) => throw_ub!("Value: slices cannot be represented as values"),
+            _ => throw_ub!("Value: value does not match type")
+        }
+
+        ret(())
+    }
+}
+```
+
+- TODO: Do we really want to special case references to uninhabited types? Do we somehow want to require more, like pointing to a valid instance of the pointee type?
+  (The latter would not even be possible with the current pointee information in MiniRust.)
+  Also see [this discussion](https://github.com/rust-lang/unsafe-code-guidelines/issues/77).
+- TODO: Inhabitedness and non-nullness are (the only two) conditions that could be checked in `decode` itself. Should we?
```

Comment:
```suggestion
- TODO: Inhabitedness, alignment, and non-nullness could be checked in `decode` itself. Should we?
```

Reply:
I removed alignment there, because it can only be checked for sized pointees.